### PR TITLE
Improve initial setup with new App Router TypeScript project

### PR DESCRIPTION
**Pull request by unstubbable**
### What?

Fixes a couple of issues in the latest Canary regarding the `tsconfig.json` adjustments that are done after creating a new App Router project with `create-next-app`. Specifically:

- Set default TS target to ES2017 (continuation of #64508).
- Do not warn about disabled strict mode if strict mode was enabled. This is the case after starting the `dev` command for the first time after having created a new project with `create-next-app`, because in the tsconfig templates `strict` is enabled per default.
- Fix wrongly printed tsconfig changes. See snapshot changes in https://github.com/vercel/next.js/commit/dc87ebcd561ef522e74fd11c0bf12356472f0268 for details of what was wrong. The bug was introduced in #64508.

### Why?

Avoid creating any changes to `tsconfig.json` in a fresh project on first run, and don't confuse the user with wrong logs.

### How?

I fixed the issues in `writeConfigurationDefaults()`, and added unit tests, as well as adjusted the existing integration tests, to prevent regressions. Additionally, I've created a new project inside of the Next.js monorepo with `pnpm create-next-app` and ran the `dev` command inside of this project to verify my fixes.

### CODE REVIEW

**Comment by ijjk** [comment]
Allow CI Workflow Run

- [ ] approve CI run for commit: 302646a0341707cc41f6023f0561a4778fe3a4c4

Note: this should only be enabled once the PR is ready to go and can only be enabled by a maintainer

**Review by eps1lon — DISMISSED** [review]
Looks good, thank you. CI just needs fixing (or probably just retrying).

**Comment by ijjk** [comment]
## Failing test suites 
Commit: da338565d2d5c8d09d0847ace79b39bbac0a34e1

`pnpm test-start test/production/jest/index.test.ts`
- next/jest > should work

<details>
<summary>Expand output</summary>

  ● next/jest › should work

    next build failed with code/signal 1

      89 |           if (code || signal)
      90 |             reject(
    > 91 |               new Error(`next build failed with code/signal ${code || signal}`)
         |               ^
      92 |             )
      93 |           else resolve()
      94 |         })

      at ChildProcess.<anonymous> (lib/next-modes/next-start.ts:91:15)


  ● Test suite failed to run

    TypeError: Cannot read properties of undefined (reading 'destroy')

      144 |     })
      145 |   })
    > 146 |   afterAll(() => next.destroy())
          |                       ^
      147 |
      148 |   it('should work', async () => {
      149 |     const html = await renderViaHTTP(next.url, '/')

      at Object.destroy (production/jest/index.test.ts:146:23)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| buildDuration | 15.7s | 16s | ⚠️ +223ms |
| buildDurationCached | 8.8s | 7.3s | N/A |
| nodeModulesSize | 199 MB | 199 MB | ⚠️ +1.16 kB |
| nextStartRea..uration (ms) | 429ms | 432ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| 2453-HASH.js gzip | 31.5 kB | 31.5 kB | N/A |
| 3304.HASH.js gzip | 169 B | 169 B | ✓ |
| 3f784ff6-HASH.js gzip | 53.7 kB | 53.7 kB | N/A |
| 8299-HASH.js gzip | 5.1 kB | 5.1 kB | N/A |
| framework-HASH.js gzip | 45.2 kB | 45.2 kB | ✓ |
| main-app-HASH.js gzip | 230 B | 228 B | N/A |
| main-HASH.js gzip | 29.6 kB | 29.7 kB | N/A |
| webpack-HASH.js gzip | 1.64 kB | 1.65 kB | N/A |
| Overall change | 45.4 kB | 45.4 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 194 B | N/A |
| _error-HASH.js gzip | 193 B | 191 B | N/A |
| amp-HASH.js gzip | 511 B | 511 B | ✓ |
| css-HASH.js gzip | 342 B | 343 B | N/A |
| dynamic-HASH.js gzip | 2.51 kB | 2.51 kB | N/A |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 365 B | 364 B | N/A |
| hooks-HASH.js gzip | 389 B | 391 B | N/A |
| image-HASH.js gzip | 4.28 kB | 4.28 kB | N/A |
| index-HASH.js gzip | 269 B | 268 B | N/A |
| link-HASH.js gzip | 2.68 kB | 2.69 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 326 B | N/A |
| script-HASH.js gzip | 395 B | 397 B | N/A |
| withRouter-HASH.js gzip | 323 B | 323 B | ✓ |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.21 kB | 1.21 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| _buildManifest.js gzip | 483 B | 485 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| index.html gzip | 528 B | 528 B | ✓ |
| link.html gzip | 541 B | 542 B | N/A |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.05 kB | 1.05 kB | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| edge-ssr.js gzip | 94.5 kB | 94.5 kB | N/A |
| page.js gzip | 3.05 kB | 3.04 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 622 B | 627 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 25.6 kB | 25.6 kB | N/A |
| edge-runtime..pack.js gzip | 839 B | 839 B | ✓ |
| Overall change | 839 B | 839 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| app-page-exp...dev.js gzip | 171 kB | 171 kB | ✓ |
| app-page-exp..prod.js gzip | 97.6 kB | 97.6 kB | ✓ |
| app-page-tur..prod.js gzip | 99.4 kB | 99.4 kB | ✓ |
| app-page-tur..prod.js gzip | 93.6 kB | 93.6 kB | ✓ |
| app-page.run...dev.js gzip | 145 kB | 145 kB | ✓ |
| app-page.run..prod.js gzip | 92.1 kB | 92.1 kB | ✓ |
| app-route-ex...dev.js gzip | 21.5 kB | 21.5 kB | ✓ |
| app-route-ex..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 15.1 kB | 15.1 kB | ✓ |
| app-route-tu..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| app-route.ru...dev.js gzip | 21.2 kB | 21.2 kB | ✓ |
| app-route.ru..prod.js gzip | 14.9 kB | 14.9 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-api.ru...dev.js gzip | 9.82 kB | 9.82 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.55 kB | 9.55 kB | ✓ |
| pages-turbo...prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| pages.runtim...dev.js gzip | 22.1 kB | 22.1 kB | ✓ |
| pages.runtim..prod.js gzip | 21.4 kB | 21.4 kB | ✓ |
| server.runti..prod.js gzip | 51.6 kB | 51.6 kB | ✓ |
| Overall change | 946 kB | 946 kB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | unstubbable/next.js improve-app-router-ts-setup | Change |
| - | - | - | - |
| 0.pack gzip | 1.6 MB | 1.6 MB | N/A |
| index.pack gzip | 107 kB | 107 kB | ⚠️ +310 B |
| Overall change | 107 kB | 107 kB | ⚠️ +310 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>middleware.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>image-HASH.js</strong></summary>

```diff
@@ -1,7 +1,7 @@
 (self["webpackChunk_N_E"] = self["webpackChunk_N_E"] || []).push([
   [8358],
   {
-    /***/ 1552: /***/ (
+    /***/ 4070: /***/ (
       __unused_webpack_module,
       __unused_webpack_exports,
       __webpack_require__
@@ -9,7 +9,7 @@
       (window.__NEXT_P = window.__NEXT_P || []).push([
         "/image",
         function () {
-          return __webpack_require__(5237);
+          return __webpack_require__(396);
         },
       ]);
       if (false) {
@@ -18,7 +18,7 @@
       /***/
     },
 
-    /***/ 2016: /***/ (module, exports, __webpack_require__) => {
+    /***/ 8490: /***/ (module, exports, __webpack_require__) => {
       "use strict";
       /* __next_internal_client_entry_do_not_use__  cjs */
       Object.defineProperty(exports, "__esModule", {
@@ -40,15 +40,15 @@
         __webpack_require__(422)
       );
       const _head = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6074)
+        __webpack_require__(2457)
       );
-      const _getimgprops = __webpack_require__(9571);
-      const _imageconfig = __webpack_require__(6567);
-      const _imageconfigcontextsharedruntime = __webpack_require__(419);
-      const _warnonce = __webpack_require__(4486);
-      const _routercontextsharedruntime = __webpack_require__(162);
+      const _getimgprops = __webpack_require__(7932);
+      const _imageconfig = __webpack_require__(5706);
+      const _imageconfigcontextsharedruntime = __webpack_require__(9483);
+      const _warnonce = __webpack_require__(9035);
+      const _routercontextsharedruntime = __webpack_require__(4829);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6996)
+        __webpack_require__(7240)
       );
       // This is replaced by webpack define plugin
       const configEnv = {
@@ -379,7 +379,7 @@
       /***/
     },
 
-    /***/ 9571: /***/ (
+    /***/ 7932: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -395,9 +395,9 @@
           return getImgProps;
         },
       });
-      const _warnonce = __webpack_require__(4486);
-      const _imageblursvg = __webpack_require__(133);
-      const _imageconfig = __webpack_require__(6567);
+      const _warnonce = __webpack_require__(9035);
+      const _imageblursvg = __webpack_require__(2642);
+      const _imageconfig = __webpack_require__(5706);
       const VALID_LOADING_VALUES =
         /* unused pure expression or super */ null && [
           "lazy",
@@ -772,7 +772,7 @@
       /***/
     },
 
-    /***/ 133: /***/ (__unused_webpack_module, exports) => {
+    /***/ 2642: /***/ (__unused_webpack_module, exports) => {
       "use strict";
       /**
        * A shared function, used on both client and server, to generate a SVG blur placeholder.
@@ -827,7 +827,7 @@
       /***/
     },
 
-    /***/ 4085: /***/ (
+    /***/ 503: /***/ (
       __unused_webpack_module,
       exports,
       __webpack_require__
@@ -854,10 +854,10 @@
         },
       });
       const _interop_require_default = __webpack_require__(2430);
-      const _getimgprops = __webpack_require__(9571);
-      const _imagecomponent = __webpack_require__(2016);
+      const _getimgprops = __webpack_require__(7932);
+      const _imagecomponent = __webpack_require__(8490);
       const _imageloader = /*#__PURE__*/ _interop_require_default._(
-        __webpack_require__(6996)
+        __webpack_require__(7240)
       );
       function getImageProps(imgProps) {
         const { props } = (0, _getimgprops.getImgProps)(imgProps, {
@@ -889,7 +889,7 @@
       /***/
     },
 
-    /***/ 6996: /***/ (__unused_webpack_module, exports) => {
+    /***/ 7240: /***/ (__unused_webpack_module, exports) => {
       "use strict";
 
       Object.defineProperty(exports, "__esModule", {
@@ -924,7 +924,7 @@
       /***/
     },
 
-    /***/ 5237: /***/ (
+    /***/ 396: /***/ (
       __unused_webpack_module,
       __webpack_exports__,
       __webpack_require__
@@ -941,8 +941,8 @@
 
       // EXTERNAL MODULE: ./node_modules/.pnpm/react@18.2.0/node_modules/react/jsx-runtime.js
       var jsx_runtime = __webpack_require__(1527);
-      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+main-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
-      var next_image = __webpack_require__(1577);
+      // EXTERNAL MODULE: ./node_modules/.pnpm/file+..+diff-repo+packages+next+next-packed.tgz_react-dom@18.2.0_react@18.2.0/node_modules/next/image.js
+      var next_image = __webpack_require__(73);
       var image_default = /*#__PURE__*/ __webpack_require__.n(next_image); // CONCATENATED MODULE: ./pages/nextjs.png
       /* harmony default export */ const nextjs = {
         src: "/_next/static/media/nextjs.cae0b805.png",
@@ -972,12 +972,8 @@
       /***/
     },
 
-    /***/ 1577: /***/ (
-      module,
-      __unused_webpack_exports,
-      __webpack_require__
-    ) => {
-      module.exports = __webpack_require__(4085);
+    /***/ 73: /***/ (module, __unused_webpack_exports, __webpack_require__) => {
+      module.exports = __webpack_require__(503);
 
       /***/
     },
@@ -987,7 +983,7 @@
     /******/ var __webpack_exec__ = (moduleId) =>
       __webpack_require__((__webpack_require__.s = moduleId));
     /******/ __webpack_require__.O(0, [2888, 9774, 179], () =>
-      __webpack_exec__(1552)
+      __webpack_exec__(4070)
     );
     /******/ var __webpack_exports__ = __webpack_require__.O();
     /******/ _N_E = __webpack_exports__;
```
</details>
<details>
<summary>Diff for <strong>2453-HASH.js</strong></summary>

Diff too large to display
</details>
<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: da338565d2d5c8d09d0847ace79b39bbac0a34e1

**Comment by ryan-nauman** [comment]
fyi this slightly relates to https://github.com/vercel/next.js/pull/61413 as well

**Comment by eps1lon** [comment]
> fyi this slightly relates to https://github.com/vercel/next.js/pull/61413 as well

@ryan-nauman How are these related?

**Comment by ryan-nauman** [comment]
@eps1lon both have tests for `writeConfigurationDefaults.ts`

* the location in this PR is `packages/next/src/lib/typescript/writeConfigurationDefaults.test.ts`
* the location in that PR is `test/unit/write-configuration-defaults.test.ts`
  
it could be confusing to have them in both places

**Comment by unstubbable** [comment]
> @eps1lon both have tests for `writeConfigurationDefaults.ts`
> 
> * the location in this PR is `packages/next/src/lib/typescript/writeConfigurationDefaults.test.ts`
> * the location in that PR is `test/unit/write-configuration-defaults.test.ts`
> 
> it could be confusing to have them in both places

Yeah, I need to merge those now, after rebasing. I'm not sure about the history of `test/unit`, but it seems that nowadays unit tests are often colocated with their implementation. The unit test in `test/unit` is also importing its implementation from `dist`, and I don't see a `moduleNameMapper`, so does this require a build when the implementation is changed? (Can't test it myself right now, but might be able to do so this evening, CET.)

**Comment by ryan-nauman** [comment]
> so does this require a build when the implementation is changed?

yes. same as `test/unit/write-app-declarations.test.ts`

i followed the guide in https://github.com/vercel/next.js/blob/canary/contributing/core/testing.md. i have one console with `pnpm dev` running. as i make changes to the implementation i run `pnpm testonly-dev ...`/`pnpm testonly-start ...`

File: packages/create-next-app/helpers/git.ts [hunk 1]
```diff
@@ -30,24 +30,26 @@ function isDefaultBranchSet(): boolean {
 export function tryGitInit(root: string): boolean {
   let didInit = false
   try {
-    execSync('git --version', { stdio: 'ignore' })
+    execSync('git --version', { stdio: 'inherit' })
     if (isInGitRepository() || isInMercurialRepository()) {
+      console.error('Not creating a git reporitory.')
```
Comment (eps1lon) [inline packages/create-next-app/helpers/git.ts:35]: Is this an error? Seems like it should just be an info here.

File: packages/create-next-app/helpers/git.ts [hunk 2]
```diff
@@ -30,24 +30,26 @@ function isDefaultBranchSet(): boolean {
...
 
-    execSync('git add -A', { stdio: 'ignore' })
+    execSync('git add -A', { stdio: 'inherit' })
     execSync('git commit -m "Initial commit from Create Next App"', {
-      stdio: 'ignore',
+      stdio: 'inherit',
     })
     return true
   } catch (e) {
+    console.error('Could not create a git reporitory.', e)
```
Comment (eps1lon) [inline packages/create-next-app/helpers/git.ts:52]: Same here. Is this really an error?

File: packages/create-next-app/helpers/git.ts [hunk 3]
```diff
@@ -30,24 +30,26 @@ function isDefaultBranchSet(): boolean {
 export function tryGitInit(root: string): boolean {
   let didInit = false
   try {
-    execSync('git --version', { stdio: 'ignore' })
+    execSync('git --version', { stdio: 'inherit' })
```
Comment (eps1lon) [inline packages/create-next-app/helpers/git.ts:33]: I feel like we omitted this on purpose. We should have access to those but not with the default log level. Do we have a configurable log level?

**Review by eps1lon — COMMENTED** [review]
Could you also update all the test fixtures with a tsconfig? Otherwise running tests in those adds unrelated changes in git.

File: packages/create-next-app/helpers/git.ts (hunk 3 above)
Comment (unstubbable) [reply packages/create-next-app/helpers/git.ts:33]: All these changes are only temporary and part of the [Trying to find out why there's not git repo created in the CI](https://github.com/vercel/next.js/pull/64826/commits/c19acba4bb096329137034b86f2715ad28adb338) commit. I will revert them when I've fixed the git CI issue.

Comment (unstubbable) [reply packages/create-next-app/helpers/git.ts:33]: Will convert to a draft until then...

**Comment by unstubbable** [comment]
> Could you also update all the test fixtures with a tsconfig? Otherwise running tests in those adds unrelated changes in git.

Do you mean "running tests" as in `node packages/next/dist/bin/next dev test/e2e/app-dir/modularizeimports`, for example? So I should update all tsconfigs in `examples` and in `test`?

**Comment by eps1lon** [comment]
> So I should update all tsconfigs in examples and in test?

At least `test`, yes. Sometimes we run the Next cli directly on those packages which would now change the tsconfig.

**Comment by unstubbable** [comment]
> > So I should update all tsconfigs in examples and in test?
> 
> At least `test`, yes. Sometimes we run the Next cli directly on those packages which would now change the tsconfig.

Alright, done.

Just to make sure we're having the same understanding regarding the statement "which would **_now_** change the tsconfig":
This PR does not change anything about that. Potentially changing the tsconfig when running `next dev` or `next lint`, is already happening on `canary`. That this currently happens for newly created projects was caused by #64508.

File: .github/workflows/build_reusable.yml [hunk 4]
```diff
@@ -133,6 +133,11 @@ jobs:
       # clean up any previous artifacts to avoid hitting disk space limits
       - run: git clean -xdf && rm -rf /tmp/next-repo-*; rm -rf /tmp/next-install-* /tmp/yarn-* /tmp/ncc-cache target
 
+      - name: Set CI git user for CNA
```
Comment (eps1lon) [inline .github/workflows/build_reusable.yml:136]: CVN numbering authority? Why is this needed now?

Comment (unstubbable) [reply .github/workflows/build_reusable.yml:136]: "Create Next App", maybe I should spell it out?

This is needed for the added integration test expectation so check whether there are any git changes after creating a new project and running `next dev`. Without this, the `git init` silently fails because no git user is defined.

Comment (eps1lon) [reply .github/workflows/build_reusable.yml]: Always good practice to spell out abbreviations on their first usage.

Comment (unstubbable) [reply .github/workflows/build_reusable.yml]: Agreed, fixed it by adding spelling it out, and adding a comment above.

**Review by eps1lon — APPROVED** [review]
I'll babysit CI to get this in. Great work, thank you!
